### PR TITLE
Refactor and Update gpdb master pipeline

### PR DESCRIPTION
**Pull request by kmacoskey**
Ok, so there are a number of things going on here. We attempted to keep the commit history very atomic so as to show each individual change we have made. There have been a number of conversations in the Greenplum Developers mailing list that relates to these proposed changes (refer to `[Concourse jobs prefixed by team name`) as well as conversation in slack about currently confusing flow of jobs in the pipeline and what jobs should block other jobs. We may not have captured every proposed change and we are happy to add more to this PR. 

**tl;dr;** of the changes:
1. Gates have been dropped unless they served an explicit purpose. 
2. Names have been updated for various things, such as DPM and MM being renamed
3. The groups of tests are now only blocked on successful compilation of the platform(s) required by the group of tests.
4. A new CL group and some other small additions to the remaining groups to make them more useful, such as making sure all jobs related to the group are included in the group.
5. DP and CM have been merged into a CLI group

Here is what the pipeline would look like with these proposed changes: https://gpdb-dev.bosh.pivotalci.info/teams/main/pipelines/dev:156173494-group-platform-gpdbmaster?groups=all

If any of the changes or intentions aren't clear in the commits/dev pipeline then please ask and we shall provide more detail.

Pinging folks that have talked about the subject:
@ashwinstar @khuddlefish @chrishajas @Eulerizeit @pivotal-mike @jvigil-pivotal @xinzweb 

Another related conversation that could happen here is whether any such changes get back-ported to `5X_STABLE`.

### CODE REVIEW

**Comment by Eulerizeit** [comment]
I like this a lot, moving to components names is super important since the names of the components won't change even if the owners do. 

Response based on a compare and contrast of the pipelines not looking at code. 

* Not an issue with/for this PR but we still have team names associated with the suites of tests. Given the volatility of team names I think we ought to remove those 
* interconnect doesn't seem to go with Resource groups wrt functionality
* MU_check_centos isn't a command line tool nor is gppkg,  I don't think, there might be some other things in that long list that belong elsewhere as well. someone from those teams maybe ought to look to see if the chunking up makes sense.   OR given there are only 2 of them that might be the best place for them.

**Comment by kmacoskey** [comment]
@jchampio Thankfully, nothing performance related is driving anything right now. One of the gates has other things driving why they it is there, but it's reasonable that others could go away.

Remaining Gates:
* `gate_compile_start` - First job to run in the pipeline. I sort of like this gate as a nice, single point of entry into starting the pipeline. But it serves to specific technical reason so I could agree to it being dropped.
* `gate_icw_end` - An artifact is currently exported here that is used in other pipelines. This may change when there is a more explicit `release_candidate` artifact exported. There may be 
arguments that it will still be desirable to only test core gpdb release candidates (passing ICW). 
* `gate_cs_start, gate_mpp_start, gate_cli_start, gate_ud_start` - Only sit in front of the group of jobs, no other specific purpose then grouping. I think the question here may be whether they continue to serve a purpose for visibility and/or being able to quickly pause an entire group at a time.

**Comment by kmacoskey** [comment]
@Eulerizeit 
> interconnect doesn't seem to go with Resource groups wrt functionality

Splitting the `MPP` group into those two functionality driven groups makes sense to me. +1

**Comment by kmacoskey** [comment]
`MPP` has been split into `Interconnect` and `ResourceGroups`. The dev pipeline above is up to date with this change.

**Comment by ashwinstar** [comment]
Good work, very helpful (yet to review in details).

Wish to discuss one more aspect as part of this PR, Is it possible we start running gpdb_master_without_asserts as well at same frequency as gpdb_master_with_asserts per commit and remove the complexity of reduced_frequency trigger. Given majority of jobs run on concourse containers shouldn't much affect the cost.

**Comment by ashwinstar** [comment]
Is `sleep_before_starting` still required ?

**Comment by kmacoskey** [comment]
> Given majority of jobs run on concourse containers shouldn't much affect the cost.

16 of the jobs are using CCP so it's not an insubstantial change in workload to run those extra jobs on every commit.

**Comment by kmacoskey** [comment]
> Is sleep_before_starting still required ?

It is not, I think it can be safely dropped.

File: concourse/pipelines/gen_pipeline.py [hunk 1]
```diff
@@ -197,7 +191,7 @@ def how_to_use_generated_pipeline_message():
     PARSER.add_argument('-a', '--test_sections',
                         action='store',
                         dest='test_sections',
-                        choices=['ICW', 'CS', 'MPP', 'MM', 'DPM', 'UD'],
+                        choices=['ICW', 'CS', 'RES', 'INT', 'CLI', 'UD'],
```
Comment (ashwinstar) [inline concourse/pipelines/gen_pipeline.py]: Reading this and even with all the Greenplum context I don't quite understand what `INT` stands for here ?

File: concourse/pipelines/gpdb_master-generated.yml [hunk 2]
```diff
@@ -116,86 +109,99 @@ groups:
...
-  - icw_gporca_conan_ubuntu16
+  - compile_gpdb_centos7
   - icw_gporca_sles11
   - icw_gporca_sles12
   - icw_planner_sles12
+  - compile_gpdb_sles11
+  - icw_planner_ubuntu16
+  - icw_gporca_conan_ubuntu16
+  - compile_gpdb_ubuntu16
   - test_gpdb_debian_package
```
Comment (ashwinstar) [inline concourse/pipelines/gpdb_master-generated.yml]: rename to ubuntu maybe ?

**Comment by pivotal-mike** [comment]
@kmacoskey minor ask - can we remove the `gate_compile_start`?

File: concourse/pipelines/gen_pipeline.py (hunk 1 above)
Comment (kmacoskey) [reply concourse/pipelines/gen_pipeline.py]: Yeah I wasn't quite sure how to abbreviate `Interconnect` or `ResourceGroups`. Looking at it now I don't see any reason it needs to a short acronym. I shall update that.

File: concourse/pipelines/gpdb_master-generated.yml (hunk 2 above)
Comment (kmacoskey) [reply concourse/pipelines/gpdb_master-generated.yml]: If we follow the convention from the release pipelines then it should be named `gpdb_packaging_ubuntu16`

**Comment by kmacoskey** [comment]
> minor ask - can we remove the gate_compile_start?

That makes +2, it shall be removed.

**Comment by kmacoskey** [comment]
If you treat `` as a single syllable then the this makes a Haiku:
`test_gpdb_debian_package` job name now fits
new acronyms are no more
`gate_compile_start` makes one less gate

I apologies for the PR being confusing in the commit/comment time line. I rebased when I probably could have waited. No way to fix that after the fact. To make up for that, all of the above has been addressed with the exception of:

> MU_check_centos isn't a command line tool nor is gppkg, I don't think, there might be some other things in that long list that belong elsewhere as well. someone from those teams maybe ought to look to see if the chunking up makes sense. OR given there are only 2 of them that might be the best place for them.

> Not an issue with/for this PR but we still have team names associated with the suites of tests. Given the volatility of team names I think we ought to remove those

> Another related conversation that could happen here is whether any such changes get back-ported to 5X_STABLE.

**Comment by ashwinstar** [comment]
On Mon, Apr 9, 2018 at 11:15 AM, Kristofer Macoskey <
notifications@github.com> wrote:

> Given majority of jobs run on concourse containers shouldn't much affect
> the cost.
>
> 16 of the jobs are using CCP so it's not an insubstantial change in
> workload to run those extra jobs on every commit.
>
Just curious to know how many of those are on nightly trigger ?

**Comment by kmacoskey** [comment]
> Just curious to know how many of those are on nightly trigger ?

Only one job, which is the job that is paused `DPM_gptransfer_43x_to_master`

File: concourse/pipelines/gpdb_master-generated.yml [hunk 3]
```diff
@@ -2379,17 +2138,17 @@ jobs:
...
+    - CM_gpinitsystem
+    - CM_gpinitstandby
+    - CM_gpexpand_1
+    - CM_gpexpand_2
+    - CM_gpcheck
+    - CM_gppkg
     - MU_check_centos
-    - DPM_gptransfer_43x_to_master
-    - DPM_gptransfer
+    - DP_gptransfer_43x_to_master
```
Comment (chrishajas) [inline concourse/pipelines/gpdb_master-generated.yml]: The gptransfer jobs do not belong to the DP team. We should either drop the DP prefix or change it to UNITE

**Comment by khuddlefish** [comment]
+1 to removing all team prefixes from jobs.

**Comment by Eulerizeit** [comment]
+ another one on removing the team prefixes. Not sure it should be this PR. I really just want this committed.

**Comment by khuddlefish** [comment]
I would strongly prefer we at least remove the prefixes from the gptransfer jobs now. I would be fine with doing a separate PR for the other prefixes if we want to have more discussion around that topic. The DP prefixes on those jobs at the moment are actively misleading and the jobs continue to be misattributed to our team as a result. If we are going to modify them, we shouldn't push something that is an incorrect modification. We can always add back a different prefix later if a future discussion leads to a new job naming standard.

**Comment by Eulerizeit** [comment]
On Wed, Apr 11, 2018 at 7:35 AM, Kristofer Macoskey <
notifications@github.com> wrote:

> Ok, there have been lots of +1 so i've dropped team prefixes and done some
> general renaming. I think it's worth trying this for now and we can always
> revisit if it's not liked. Dev pipeline is up to date and viewable here:
> https://gpdb-dev.bosh.pivotalci.info/teams/main/
> pipelines/dev:156173494-group-platform-gpdbmaster?groups=all
>
> The only outstanding topics were:
>
> MU_check_centos isn't a command line tool nor is gppkg, I don't think,
> there might be some other things in that long list that belong elsewhere as
> well. someone from those teams maybe ought to look to see if the chunking
> up makes sense. OR given there are only 2 of them that might be the best
> place for them.
>
> I think this can be punted to a different PR since there is already a lot
> going on here unless someone would like to weigh in now.
>
> Another related conversation that could happen here is whether any such
> changes get back-ported to 5X_STABLE.
>
> What is the advantage of backporting this? Seems like the 5.X CI is
relatively stable and contains many many more TINC tests which could stress
the CI. In general I'm -1 unless there is a good reason.

**Comment by skahler-pivotal** [comment]
I'm going to +1 the backport. My view is that there are a lot of features still being backported to 5.x ( which is a different discussion ) and as long as we are in that state maintaining similarity between master and 5x reduces our chances of creating an issue while doing that backport. Worth re-discussion in the future.

**Comment by Eulerizeit** [comment]
> I'm going to +1 the backport. My view is that there are a lot of features still being backported to 5.x ( which is a different discussion ) and as long as we are in that state maintaining similarity between master and 5x reduces our chances of creating an issue while doing that backport. Worth re-discussion in the future.

I don't understand how reorganizing the CI and changing the names will make it easier to backport things to 5.x. What I do see is us removing a lot of the things we have added for CI stability so that the master pipeline has a more logical blocking structure etc. 

As I see it backporting this is just asking for potential CI instability for the sake of aesthetics.

**Comment by jmcatamney** [comment]
Regarding check_centos/gppkg: I think they make sense where they are and don't need to be punted. check_centos should definitely stay where it is, since it's running the unit tests for Python utilities so splitting it out of the CLI group doesn't make sense, and while gppkg isn't as strongly associated with the other utilities it makes the most sense there because it shares all the gppylib code with other Python utilities and so a change that affects them would be likely to affect gppkg as well.

Regarding backporting to 5.X: I agree with @skahler-pivotal that until we stop backporting lots of master commits, keeping the pipelines in sync makes that process easier and less error-prone.  +1 to the backport.

**Comment by chrishajas** [comment]
I would also +1 to backport if it's not too much work and won't burden concourse resources too much. It seems like we typically end up backporting concourse changes more often than we'd like for one reason or another (whether just restructuring, test refactoring, release process, etc.) and keeping the same pipeline structure makes this easier.

**Comment by slari-pivotal** [comment]
+1 for backport. Having the same pipeline structure reduces the chances of not running a test suite for a backport.

**Comment by Eulerizeit** [comment]
The thing about this I DO NOT want backported is the removal of the staggered starts that the gates have added. I don't care bout the naming or grouping that is fine. Can we please NOT introduce potential instability in the pipeline?

**Comment by jmcatamney** [comment]
If the main concern is all the extra TiNC tests in the 5.X pipeline, would it make sense/be possible to leave timing gates around those tests while making the rest of the pipeline match the master one?

**Pull request merged**